users/vieting/experiments/switchboard/ctc/feat/experiments.py

```diff
@@ -719,41 +724,354 @@ def run_scf_baseline():
+        {"prob": 0.6, "minimum": 0.8, "maximum": 1.2},
+        {"prob": 0.6, "minimum": 0.7, "maximum": 1.3},
+        {"prob": 0.5, "minimum": 0.9, "maximum": 1.1},
+        {"prob": 0.5, "minimum": 0.8, "maximum": 1.2},
+        {"prob": 0.5, "minimum": 0.7, "maximum": 1.3},
+        {"prob": 0.4, "minimum": 0.9, "maximum": 1.1},
+        {"prob": 0.4, "minimum": 0.8, "maximum": 1.2},
+        {"prob": 0.4, "minimum": 0.7, "maximum": 1.3},
+    ]
+
+    tempos = [
+        {"prob": 0.4, "minimum": 0.9, "maximum": 1.1},
+        {"prob": 0.4, "minimum": 0.8, "maximum": 1.2},
+        {"prob": 0.4, "minimum": 0.7, "maximum": 1.3},
+        {"prob": 0.5, "minimum": 0.9, "maximum": 1.1},
+        {"prob": 0.5, "minimum": 0.8, "maximum": 1.2},
+        {"prob": 0.5, "minimum": 0.7, "maximum": 1.3},
+        {"prob": 0.6, "minimum": 0.9, "maximum": 1.1},
+        {"prob": 0.6, "minimum": 0.8, "maximum": 1.2},
+        {"prob": 0.6, "minimum": 0.7, "maximum": 1.3},
+    ]
+    preemphases = [
+        {"prob": 0.9, "minimum": 0.9, "maximum": 1.0},
+        {"prob": 0.9, "minimum": 0.8, "maximum": 1.0},
+        {"prob": 0.8, "minimum": 0.9, "maximum": 1.0},
+        {"prob": 0.8, "minimum": 0.8, "maximum": 1.0},
+    ]
+    codecs = [
+        {"encoding": "ULAW", "prob": 0.4},
+        {"encoding": "ULAW", "prob": 0.6},
+    ]
+    non_linearities = [
+        {"prob": 0.4, "alpha": 0.2},
+        {"prob": 0.4, "alpha": 0.4},
+        {"prob": 0.6, "alpha": 0.2},
+        {"prob": 0.6, "alpha": 0.4},
+    ]
+
+    nn_base_args = {}
+
+    for speed in speeds:
+        key_suffix = []
+        # build key suffix
+        key_suffix.append("speed" + f"{speed['prob']}_{speed['minimum']}_{speed['maximum']}")
+        key = f"scf_bs7k_conf-wei-oldspecaug-audio_perturbation_{'_'.join(key_suffix)}"
+        audio_perturb_args = {}
+        audio_perturb_args["speed"] = speed
+        report_args = {
+            "architecture": "conf-wei",
+            "lr": "wei_peak_4e-4_e450_cycle360",
+            "specaug": "wei_adapt_80dim",
+            "wave_norm": "True",
+        }
+        report_args["speed"] = f"{speed['prob']}_{speed['minimum']}_{speed['maximum']}"
+        nn_base_args[key] = dict(
+            returnn_args={
+                "extra_args": {
+                    "audio_perturb_args": audio_perturb_args,
+                    "audio_perturb_runner": CodeWrapper("WaveformPerturbation(**audio_perturb_args)"),
+                    "conv_pad_seq_len_to_power": 1.5,
+                    "watch_memory": True,
                 },
-                report_args={
-                    "architecture": "conf-wei",
-                    "lr": "wei_peak_4e-4_e450_cycle360",
-                    "specaug": "wei_adapt_80dim",
-                    "wave_norm": "True",
+                **returnn_args,
+            },
+            feature_args=feature_args,
+            lr_args=lr_args,
+            report_args=report_args,
+        )
+
+    for tempo in tempos:
```

Comment:
Please check the comments I added for speed perturbation for all perturbations below. Also, there is a lot of code duplication below. I think you could solve this much more elegantly for the different perturbations.

Reply:
@Max-Ryujin This is still valid. Can't you e.g. create a list with all perturbations and then have just one loop through it? Something like
```
perturbations = [{"speed": [<list that you now have for speed>], "tempo": ...]
for perturbation in perturbations:
    ...
```
Then you can use `perturbation` instead of `audio_perturb_args` and just need a slight generalization of how you create the name.

Reply:
That would make the code shorter, but maybe less understandable. There are also differences in the amount and names of parameters each perturbation needs, which would make the `WaveformPerturbation` call more complex

Reply:
The code for each type of perturbation is exactly the same except for alias, report args and creation of `audio_perturb_args`. If you create a list of all perturbations where each entry is exactly what can be passed to `audio_perturb_args`, I do not think this is overly complicated or difficult to read. You just need a way to parse name and report args from that entry.